File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/tarjan_scc.py
from __future__ import annotations


def tarjan_scc(graph: dict[str, list[str]]) -> list[list[str]]:
    index = 0
    stack: list[str] = []
    on_stack: set[str] = set()
    indices: dict[str, int] = {}
    lowlink: dict[str, int] = {}
    components: list[list[str]] = []

    def strongconnect(vertex: str) -> None:
        nonlocal index
        indices[vertex] = index
        lowlink[vertex] = index
        index += 1
        stack.append(vertex)
        on_stack.add(vertex)

        for neighbor in graph.get(vertex, []):
            if neighbor not in indices:
                strongconnect(neighbor)
                lowlink[vertex] = min(lowlink[vertex], lowlink[neighbor])
            elif neighbor in on_stack:
                lowlink[vertex] = min(lowlink[vertex], indices[neighbor])

        if lowlink[vertex] == indices[vertex]:
            component = []
            while True:
                node = stack.pop()
                on_stack.remove(node)
                component.append(node)
                if node == vertex:
                    break
            components.append(component)

    for vertex in graph:
        if vertex not in indices:
            strongconnect(vertex)
    return components


if __name__ == "__main__":
    sample = {"A": ["B"], "B": ["C"], "C": ["A", "D"], "D": []}
    print("Tarjan strongly connected components")
    print(tarjan_scc(sample))

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/TarjanScc.java
package com.example.graphalgorithms;

import java.util.*;

public class TarjanScc {
    private int index = 0;
    private final Deque<String> stack = new ArrayDeque<>();
    private final Set<String> onStack = new HashSet<>();
    private final Map<String, Integer> indices = new HashMap<>();
    private final Map<String, Integer> lowlink = new HashMap<>();
    private final List<List<String>> components = new ArrayList<>();

    public static List<List<String>> find(Map<String, List<String>> graph) {
        TarjanScc solver = new TarjanScc();
        for (String vertex : graph.keySet()) {
            if (!solver.indices.containsKey(vertex)) {
                solver.strongConnect(graph, vertex);
            }
        }
        return solver.components;
    }

    private void strongConnect(Map<String, List<String>> graph, String vertex) {
        indices.put(vertex, index);
        lowlink.put(vertex, index);
        index++;
        stack.push(vertex);
        onStack.add(vertex);

        for (String neighbor : graph.getOrDefault(vertex, List.of())) {
            if (!indices.containsKey(neighbor)) {
                strongConnect(graph, neighbor);
                lowlink.put(vertex, Math.min(lowlink.get(vertex), lowlink.get(neighbor)));
            } else if (onStack.contains(neighbor)) {
                lowlink.put(vertex, Math.min(lowlink.get(vertex), indices.get(neighbor)));
            }
        }

        if (lowlink.get(vertex).equals(indices.get(vertex))) {
            List<String> component = new ArrayList<>();
            String node;
            do {
                node = stack.pop();
                onStack.remove(node);
                component.add(node);
            } while (!node.equals(vertex));
            components.add(component);
        }
    }

    public static void main(String[] args) {
        System.out.println("Tarjan strongly connected components");
        System.out.println(find(Map.of("A", List.of("B"), "B", List.of("C"), "C", List.of("A", "D"), "D", List.of())));
    }
}

